File: frontend/src/api/roles.ts
import { request } from '@/utils/request'
import type { ApiResponse } from '@/types/common'

// 角色相关类型定义
export interface Role {
  id: number
  name: string
  code: string
  description?: string
  status: boolean
  is_system: boolean
  created_at: string
  updated_at: string
  created_by?: number
  updated_by?: number
}

export interface RoleCreate {
  name: string
  code: string
  description?: string
  status?: boolean
}

export interface RoleUpdate {
  name?: string
  code?: string
  description?: string
  status?: boolean
}

export interface UserRole {
  id: number
  user_id: number
  role_id: number
  created_at: string
  created_by?: number
}

export interface UserRoleCreate {
  user_id: number
  role_id: number
}

export interface RolePermission {
  id: number
  role_id: number
  permission_id: number
  created_at: string
  created_by?: number
}

export interface RolePermissionBatchCreate {
  role_id: number
  permission_ids: number[]
}

export interface RoleWithPermissions extends Role {
  permissions: Permission[]
}

export interface UserWithRoles {
  id: number
  username: string
  email: string
  roles: Role[]
}

export interface Permission {
  id: number
  name: string
  code: string
  description?: string
  module: string
  resource: string | null
  action: string
  status: boolean
  sort_order: number
  is_system: boolean
  created_at: string
  updated_at: string
}

// 角色分页数据结构（与后端 RoleListResponse 对齐）
export interface RoleListResponse {
  items: Role[]
  total: number
  page: number
  size: number
  pages: number
}

// 角色管理API接口
export const roleApi = {
  /**
   * 获取角色列表
   */
  getRoles(params?: {
    page?: number
    size?: number
    search?: string
    status?: boolean
    is_system?: boolean
  }) {
    return request.get<ApiResponse<RoleListResponse>>('/roles/', { params })
  },

  /**
   * 获取角色详情
   */
  getRole(roleId: number) {
    return request.get<ApiResponse<Role>>(`/roles/${roleId}`)
  },

  /**
   * 创建角色
   */
  createRole(data: RoleCreate) {
    return request.post<ApiResponse<Role>>('/roles/', data)
  },

  /**
   * 更新角色
   */
  updateRole(roleId: number, data: RoleUpdate) {
    return request.put<ApiResponse<Role>>(`/roles/${roleId}`, data)
  },

  /**
   * 删除角色
   */
  deleteRole(roleId: number) {
    return request.delete<ApiResponse<any>>(`/roles/${roleId}`)
  },

  /**
   * 获取角色权限
   */
  getRolePermissions(roleId: number) {
    return request.get<ApiResponse<RoleWithPermissions>>(`/roles/${roleId}/permissions`)
  },

  /**
   * 为角色分配权限
   */
  assignPermissions(roleId: number, permissionIds: number[]) {
    // 后端要求 Body 同时包含 role_id 与 permission_ids
    return request.post<ApiResponse<any>>(`/roles/${roleId}/permissions`, {
      role_id: roleId,
      permission_ids: permissionIds
    })
  },

  /**
   * 移除角色权限
   */
  removePermission(roleId: number, permissionId: number) {
    return request.delete<ApiResponse<any>>(`/roles/${roleId}/permissions/${permissionId}`)
  },

  /**
   * 批量移除角色权限
   */
  removePermissions(roleId: number, permissionIds: number[]) {
    return request.delete<ApiResponse<any>>(`/roles/${roleId}/permissions/batch`, {
      data: { permission_ids: permissionIds }
    })
  },

  /**
   * 获取角色用户列表
   */
  getRoleUsers(roleId: number, params?: {
    page?: number
    size?: number
  }) {
    return request.get<ApiResponse<UserWithRoles[]>>(`/roles/${roleId}/users`, { params })
  },

  /**
   * 为用户分配角色
   */
  assignRole(data: UserRoleCreate) {
    return request.post<ApiResponse<UserRole>>('/roles/assign-user', data)
  },

  /**
   * 移除用户角色
   */
  removeRole(userId: number, roleId: number) {
    return request.delete<ApiResponse<any>>(`/roles/remove-user/${userId}/${roleId}`)
  },

  /**
   * 获取用户角色列表
   */
  getUserRoles(userId: number) {
    return request.get<ApiResponse<Role[]>>(`/roles/users/${userId}/roles`)
  },

  /**
   * 批量分配角色
   */
  batchAssignRoles(userIds: number[], roleIds: number[]) {
    return request.post('/roles/batch-assign', {
      user_ids: userIds,
      role_ids: roleIds
    })
  },

  /**
   * 批量移除角色
   */
  batchRemoveRoles(userIds: number[], roleIds: number[]) {
    return request.post('/roles/batch-remove', {
      user_ids: userIds,
      role_ids: roleIds
    })
  }
}

// 权限管理API接口
export const permissionApi = {
  /**
   * 获取权限列表
   *
   * 说明：后端参数为 `skip` 与 `limit`，若传入 `page/size`，后端会忽略。
   * 为兼容现有用法，保留 `page/size` 定义，但建议在批量操作或全量拉取时使用 `skip/limit`。
   */
  getPermissions(params?: {
    page?: number
    size?: number
    skip?: number
    limit?: number
    search?: string
    module?: string
    resource?: string
    action?: string
    status?: boolean
  }) {
    return request.get<ApiResponse<{ items: Permission[]; total: number; page: number; size: number; pages: number }>>('/permissions/', { params })
  },

  /**
   * 获取权限详情
   */
  getPermission(permissionId: number) {
    return request.get<ApiResponse<Permission>>(`/permissions/${permissionId}`)
  },

  /**
   * 创建权限
   *
   * 说明：`resource` 可为 `null`，用于表示非特定资源的操作权限。
   */
  createPermission(data: {
    name: string
    code: string
    description?: string
    module: string
    resource: string | null
    action: string
    status?: boolean
    sort_order?: number
  }) {
    return request.post<ApiResponse<Permission>>('/permissions/', data)
  },

  /**
   * 更新权限
   */
  updatePermission(permissionId: number, data: {
    name?: string
    code?: string
    description?: string
    module?: string
    resource?: string
    action?: string
    status?: boolean
    sort_order?: number
  }) {
    return request.put<ApiResponse<Permission>>(`/permissions/${permissionId}`, data)
  },

  /**
   * 删除权限
   */
  deletePermission(permissionId: number) {
    return request.delete<ApiResponse<any>>(`/permissions/${permissionId}`)
  },

  /**
   * 获取权限树结构
   */
  getPermissionTree() {
    // 后端返回 ResponseModel[PermissionTreeListResponse]
    return request.get<ApiResponse<any>>('/permissions/tree')
  },

  /**
   * 根据模块获取权限
   */
  getPermissionsByModule(module: string) {
    // 后端路由为 /permissions/modules/{module}
    return request.get<ApiResponse<Permission[]>>(`/permissions/modules/${module}`)
  },

  /**
   * 批量创建权限
   * 请求体为权限数组（List[PermissionCreate]），不是对象包装
   */
  batchCreatePermissions(permissions: Array<{
    name: string
    code: string
    description?: string
    module: string
    resource: string | null
    action: string
    status?: boolean
    sort_order?: number
  }>) {
    return request.post<ApiResponse<any>>('/permissions/batch', permissions)
  },

  /**
   * 检查用户权限
   */
  checkUserPermission(userId: number, permissionCode: string) {
    return request.get<ApiResponse<{ has_permission: boolean }>>(`/permissions/check/${userId}/${permissionCode}`)
  },

  /**
   * 检查当前认证用户的权限
   * @param permissionCode 权限代码
   * @returns 是否拥有该权限（基于认证信息解析当前用户）
   */
  checkMyPermission(permissionCode: string) {
    return request.get<ApiResponse<boolean>>(`/permissions/check/${permissionCode}`)
  },

  /**
   * 获取用户所有权限
   */
  getUserPermissions(userId: number) {
    // 后端路由为 /permissions/user/{user_id}，返回 ResponseModel[UserPermissionResponse]
    return request.get<ApiResponse<any>>(`/permissions/user/${userId}`)
  },

  /**
   * 获取当前认证用户的所有权限
   * @returns 后端返回 ResponseModel[UserPermissionResponse]
   */
  getMyPermissions() {
    return request.get<ApiResponse<any>>('/permissions/me')
  }
}

File: backend/app/api/v1/endpoints/roles.py
"""
角色管理API路由
提供角色的CRUD操作接口
"""
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.api.dependencies import get_db, get_current_user_sync as get_current_user
from app.crud.role import crud_role, crud_user_role, crud_role_permission
from app.crud.permission import crud_permission
from app.models.user import User
from app.schemas.role import (
    RoleCreate, RoleUpdate, RoleResponse, RoleListResponse,
    UserRoleCreate, UserRoleResponse, 
    RolePermissionCreate, RolePermissionBatchCreate, RolePermissionResponse,
    RoleWithPermissions, UserWithRoles
)
from app.schemas.common import ResponseModel

router = APIRouter()


@router.get("/", response_model=ResponseModel[RoleListResponse])
def get_roles(
    skip: int = Query(0, ge=0, description="跳过的记录数"),
    limit: int = Query(100, ge=1, le=1000, description="返回的记录数"),
    name: Optional[str] = Query(None, description="角色名称（模糊搜索）"),
    status: Optional[bool] = Query(None, description="状态筛选"),
    is_system: Optional[bool] = Query(None, description="是否系统角色"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    获取角色列表
    支持分页和筛选
    """
    roles, total = crud_role.get_multi(
        db=db, 
        skip=skip, 
        limit=limit,
        name=name,
        status=status,
        is_system=is_system
    )
    
    return ResponseModel(
        data=RoleListResponse(
            items=[RoleResponse.model_validate(role) for role in roles],
            total=total,
            page=skip // limit + 1,
            size=limit
        ),
        message="获取角色列表成功"
    )


@router.get("/{role_id}", response_model=ResponseModel[RoleResponse])
def get_role(
    role_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    根据ID获取角色详情
    """
    role = crud_role.get(db=db, role_id=role_id)
    if not role:
        raise HTTPException(status_code=404, detail="角色不存在")
    
    return ResponseModel(
        data=RoleResponse.model_validate(role),
        message="获取角色详情成功"
    )


@router.post("/", response_model=ResponseModel[RoleResponse])
def create_role(
    role_in: RoleCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    创建新角色
    """
    # 检查角色代码是否已存在
    existing_role = crud_role.get_by_code(db=db, code=role_in.code)
    if existing_role:
        raise HTTPException(status_code=400, detail="角色代码已存在")
    
    # 检查角色名称是否已存在
    existing_name = crud_role.get_by_name(db=db, name=role_in.name)
    if existing_name:
        raise HTTPException(status_code=400, detail="角色名称已存在")
    
    role = crud_role.create(db=db, obj_in=role_in, created_by=current_user.id)
    
    return ResponseModel(
        data=RoleResponse.model_validate(role),
        message="创建角色成功"
    )


@router.put("/{role_id}", response_model=ResponseModel[RoleResponse])
def update_role(
    role_id: int,
    role_in: RoleUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    更新角色信息
    """
    role = crud_role.get(db=db, role_id=role_id)
    if not role:
        raise HTTPException(status_code=404, detail="角色不存在")
    
    # 系统角色不能修改
    if role.is_system:
        raise HTTPException(status_code=400, detail="系统角色不能修改")
    
    # 检查角色代码是否已存在（排除当前角色）
    if role_in.code and role_in.code != role.code:
        existing_role = crud_role.get_by_code(db=db, code=role_in.code)
        if existing_role:
            raise HTTPException(status_code=400, detail="角色代码已存在")
    
    # 检查角色名称是否已存在（排除当前角色）
    if role_in.name and role_in.name != role.name:
        existing_name = crud_role.get_by_name(db=db, name=role_in.name)
        if existing_name:
            raise HTTPException(status_code=400, detail="角色名称已存在")
    
    role = crud_role.update(db=db, db_obj=role, obj_in=role_in, updated_by=current_user.id)
    
    return ResponseModel(
        data=RoleResponse.model_validate(role),
        message="更新角色成功"
    )


@router.delete("/{role_id}", response_model=ResponseModel[bool])
def delete_role(
    role_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    删除角色
    """
    role = crud_role.get(db=db, role_id=role_id)
    if not role:
        raise HTTPException(status_code=404, detail="角色不存在")
    
    # 系统角色不能删除
    if role.is_system:
        raise HTTPException(status_code=400, detail="系统角色不能删除")
    
    # 检查是否有用户使用该角色
    users = crud_role.get_users_by_role(db=db, role_id=role_id)
    if users:
        raise HTTPException(status_code=400, detail=f"该角色正在被 {len(users)} 个用户使用，无法删除")
    
    success = crud_role.delete(db=db, role_id=role_id)
    if not success:
        raise HTTPException(status_code=400, detail="删除角色失败")
    
    return ResponseModel(
        data=True,
        message="删除角色成功"
    )


@router.get("/{role_id}/permissions", response_model=ResponseModel[RoleWithPermissions])
def get_role_permissions(
    role_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    获取角色的权限列表
    """
    role = crud_role.get_with_permissions(db=db, role_id=role_id)
    if not role:
        raise HTTPException(status_code=404, detail="角色不存在")
    
    return ResponseModel(
        data=RoleWithPermissions.model_validate(role),
        message="获取角色权限成功"
    )


@router.post("/{role_id}/permissions", response_model=ResponseModel[List[RolePermissionResponse]])
def assign_permissions_to_role(
    role_id: int,
    permission_data: RolePermissionBatchCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    为角色分配权限
    """
    role = crud_role.get(db=db, role_id=role_id)
    if not role:
        raise HTTPException(status_code=404, detail="角色不存在")
    
    # 系统角色不能修改权限
    if role.is_system:
        raise HTTPException(status_code=400, detail="系统角色不能修改权限")
    
    # 验证权限是否存在
    for permission_id in permission_data.permission_ids:
        permission = crud_permission.get(db=db, permission_id=permission_id)
        if not permission:
            raise HTTPException(status_code=400, detail=f"权限ID {permission_id} 不存在")
    
    role_permissions = crud_role_permission.assign_permissions_to_role(
        db=db, 
        role_id=role_id, 
        permission_ids=permission_data.permission_ids,
        created_by=current_user.id
    )
    
    return ResponseModel(
        data=[RolePermissionResponse.model_validate(rp) for rp in role_permissions],
        message="分配权限成功"
    )


@router.delete("/{role_id}/permissions/{permission_id}", response_model=ResponseModel[bool])
def remove_permission_from_role(
    role_id: int,
    permission_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    移除角色的权限
    """
    role = crud_role.get(db=db, role_id=role_id)
    if not role:
        raise HTTPException(status_code=404, detail="角色不存在")
    
    # 系统角色不能修改权限
    if role.is_system:
        raise HTTPException(status_code=400, detail="系统角色不能修改权限")
    
    success = crud_role_permission.delete(db=db, role_id=role_id, permission_id=permission_id)
    if not success:
        raise HTTPException(status_code=400, detail="移除权限失败")
    
    return ResponseModel(
        data=True,
        message="移除权限成功"
    )


@router.get("/{role_id}/users", response_model=ResponseModel[List[UserWithRoles]])
def get_role_users(
    role_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    获取拥有指定角色的用户列表
    """
    role = crud_role.get(db=db, role_id=role_id)
    if not role:
        raise HTTPException(status_code=404, detail="角色不存在")
    
    users = crud_role.get_users_by_role(db=db, role_id=role_id)
    
    return ResponseModel(
        data=[UserWithRoles.model_validate(user) for user in users],
        message="获取角色用户成功"
    )


@router.get("/users/{user_id}/roles", response_model=ResponseModel[List[RoleResponse]])
def get_user_roles(
    user_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取指定用户的角色列表（函数级注释）

    - 基于用户ID查询当前有效角色
    - 返回 `RoleResponse` 数组，便于前端进行差异同步
    """
    roles = crud_user_role.get_roles_by_user(db=db, user_id=user_id)
    return ResponseModel(
        data=[RoleResponse.model_validate(r) for r in roles],
        message="获取用户角色成功"
    )


@router.post("/assign-user", response_model=ResponseModel[UserRoleResponse])
def assign_role_to_user(
    user_role_in: UserRoleCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    为用户分配角色
    """
    # 验证用户是否存在
    from app.crud.user import crud_user
    user = crud_user.get(db=db, id=user_role_in.user_id)
    if not user:
        raise HTTPException(status_code=400, detail="用户不存在")
    
    # 验证角色是否存在
    role = crud_role.get(db=db, role_id=user_role_in.role_id)
    if not role:
        raise HTTPException(status_code=400, detail="角色不存在")
    
    user_role = crud_user_role.create(db=db, obj_in=user_role_in, created_by=current_user.id)
    
    return ResponseModel(
        data=UserRoleResponse.model_validate(user_role),
        message="分配角色成功"
    )


@router.delete("/remove-user", response_model=ResponseModel[bool])
def remove_role_from_user(
    user_id: int,
    role_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    移除用户的角色
    """
    success = crud_user_role.delete(db=db, user_id=user_id, role_id=role_id)
    if not success:
        raise HTTPException(status_code=400, detail="移除角色失败")
    
    return ResponseModel(
        data=True,
        message="移除角色成功"
    )


@router.delete("/remove-user/{user_id}/{role_id}", response_model=ResponseModel[bool])
def remove_role_from_user_path(
    user_id: int,
    role_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """移除用户的角色（函数级注释）

    - 使用路径参数明确传递 `user_id` 与 `role_id`
    - 避免与 `/{role_id}` 动态路由匹配冲突，确保稳定删除
    """
    success = crud_user_role.delete(db=db, user_id=user_id, role_id=role_id)
    if not success:
        raise HTTPException(status_code=400, detail="移除角色失败")
    return ResponseModel(data=True, message="移除角色成功")